File: Proyectos/club_ciclista_ruby/el_club_ruby/lib/bicicleta.rb
#encoding: UTF-8

module El_Club
  class Bicicleta
  
     attr_accessor  :color; # variable de instancia de la que se genera de forma automática el consultor y modificador de color y público
     attr_reader    :marcha, :NUMEROSERIE, :kilometrosRecorridos; # variables de instancia de las que se genera de forma automática el consultor de estos atributos y público 
  
    @@NumeroBicicletas=0  # Variable de ámbito de clase, compartida por la clase, el módulo y submodulos en el que está definida
    @@NUMERORUEDAS=2      # Igual que la anterior
 
    def initialize(*args) # inicializador de los atributos 
      case args.size
        when 0
          @marcha=0
          @color="gris"
        when 2
          @marcha, @color =args
      end
      @kilometrosRecorridos= 0
      Bicicleta.incrementarNumeroBicicletas;
      @NUMEROSERIE=@@NumeroBicicletas+1  
    end

  # consultores
  def self.numeroBicicletas  # Método de instancia de clase
    @@NumeroBicicletas
  end 
  
  
  def self.incrementarNumeroBicicletas # Método de clase
    @@NumeroBicicletas += 1
  end 
 
  def numeroRuedas # Método de instancia que accede a variable de ámbito de clase
    @@NUMERORUEDAS;
    
  end     
 
  def hacerRuta(lugar, km)
     desplazarse(lugar)      
     incrementarKilometros(km)   # llamando a un método private no puede usarse self
     
  end  

  def desplazarse(lugar)
   puts "Voy con mi bicicleta #{@color} hasta #{lugar}"
  end  
  
  def to_s
    "el color de mi bici es #{@color} y tiene #{@marcha} marchas"
  end
  
  private  # los siguientes metodos son privados     
   
  def incrementarKilometros(km)
       @kilometrosRecorridos += km
      
  end 
   
  end # fin de la clase
  

end # fin del módulo UnModulo

File: Proyectos/ClubCiclistaJava/ElClub/src/el_club/Bicicleta.java
// Ejemplo del tema 2.1, para la prueba de visibilidad y ámbito de los atributos y métodos de una clase

package el_club; 

public class Bicicleta {
    
    //Atributos de ámbito de clase
    private static int NumeroBicicletas = 0; 
    public static final int NUMERORUEDAS = 2; // CONSTANTE
    
    // Atributos de ámbito de instancia todos privados
    private int marchas = 5;  
    private String color;
    private double kilometrosRecorridos; 
    private final int NUMEROSERIE; //CONSTANTE, no debe cambiar
    
    
    // Métodos de ámbito de clase     
    public static int getNumeroBicicletas(){        
        return NumeroBicicletas;
    } 
    public static int getNumeroRuedas(){
        return NUMERORUEDAS;
    } 
    private static void incrementarNumeroBicicletas(){
         NumeroBicicletas++;
    }   
    
    // Constructor por defecto
    public Bicicleta(){
     	this.marchas=0;
	this.color="gris"; 
        this.kilometrosRecorridos = 0;
        NUMEROSERIE = NumeroBicicletas; 
        Bicicleta.incrementarNumeroBicicletas();
    }
    // Construtor
    Bicicleta(int marcha, String color){
        this.marchas = marcha;  
        this.color = color;
        this.kilometrosRecorridos = 0;
        NUMEROSERIE = NumeroBicicletas;  
        Bicicleta.incrementarNumeroBicicletas();
    }
   
    // Métodos de instancia: Consultores básicos
    public String getColor()  {
	  return color;
    }    
    public int getMarchas() {
          return marchas;
    }   
     
    public int getNumeroSerie(){
        return NUMEROSERIE;
    }
    double getKilometros(){
        return kilometrosRecorridos;
    }
    
    // Métodos de instancias: Modificadores 
    public void setColor(String unColor) {
	  color= unColor;
    }
    public void hacerRuta(String lugar, double km)
    {
        this.desplazarse(lugar);
        this.incrementarKilometros(km);
    }
    void desplazarse(String lugar)
    {
        System.out.println("Voy con la bicicleta hasta "+ lugar);
    }
    public void incrementarKilometros(double km){
        kilometrosRecorridos+=km;
    }
    public String toString(){
        return " color " + color + " he recorrido " + kilometrosRecorridos + " km";
    }
    
}
